crates/shared/src/price_estimation/native/oneinch.rs

```diff
@@ -0,0 +1,153 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Context, Result},
+    ethrpc::current_block::{into_stream, CurrentBlockStream},
+    futures::{future::BoxFuture, FutureExt, StreamExt},
+    primitive_types::{H160, U256},
+    reqwest::{header::AUTHORIZATION, Client, StatusCode},
+    std::{
+        collections::HashMap,
+        sync::{Arc, Mutex},
+    },
+};
+
+const BASE_URL: &str = "https://api.1inch.dev/";
+
+pub struct OneInch {
+    // Denominated in wei
+    prices: Arc<Mutex<HashMap<H160, U256>>>,
+}
+
+impl OneInch {
+    #[allow(dead_code)]
+    pub fn new(
+        client: Client,
+        api_key: Option<String>,
+        chain_id: u64,
+        current_block: CurrentBlockStream,
+    ) -> Self {
+        let instance = Self {
+            prices: Arc::new(Mutex::new(HashMap::new())),
+        };
+        instance.update_prices_in_background(client, api_key, chain_id, current_block);
+        instance
+    }
+
+    fn update_prices_in_background(
+        &self,
+        client: Client,
+        api_key: Option<String>,
+        chain_id: u64,
+        current_block: CurrentBlockStream,
+    ) {
+        let prices = self.prices.clone();
+        tokio::task::spawn(async move {
+            let mut block_stream = into_stream(current_block);
+            loop {
+                match update_prices(&client, api_key.clone(), chain_id).await {
+                    Ok(new_prices) => {
+                        tracing::debug!("OneInch spot prices updated");
+                        *prices.lock().unwrap() = new_prices;
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err, "OneInch spot price update failed");
+                    }
+                }
+                block_stream.next().await;
+            }
+        });
+    }
+}
+
+impl NativePriceEstimating for OneInch {
+    fn estimate_native_price(&self, token: H160) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let prices = self.prices.lock().unwrap();
+            let price = prices
+                .get(&token)
+                .ok_or_else(|| PriceEstimationError::NoLiquidity)?;
+            Ok(price.to_f64_lossy() / 1e18)
+        }
+        .boxed()
+    }
+}
+
+async fn update_prices(
+    client: &Client,
+    api_key: Option<String>,
+    chain: u64,
+) -> Result<HashMap<H160, U256>> {
+    let mut builder = client.get(format!("{}/price/v1.1/{}", BASE_URL, chain));
```

Comment:
Just noticed, but this should likely be customizable so that we can funnel it through our egress.